### PR TITLE
Add prereq option

### LINKED ISSUE #147: need to getting list of externals in the same order that is defined in the config file
I am trying to checkout following config file that is created for the NOAA UFS application,

```
[externals_description]
schema_version = 1.0.0

[hafs]
branch = feature/hafs_nogsi
protocol = git
repo_url = https://github.com/hafs-community/HAFS.git
local_path = HAFS
required = True

[hycom]
# feature/NUOPC_merged
tag = 28b84ce
protocol = git
repo_url = https://github.com/hafs-community/HYCOM-src.git
local_path = HAFS/sorc/hafs_forecast.fd/HYCOM
required = True

[cime]
# master 
tag = 35fb2d7
protocol = git
repo_url = https://github.com/ESMCI/cime.git
local_path = cime
externals = ../Externals_cime.cfg
required = True

[cdeps]
# master 
tag = e99ec4f
protocol = git
repo_url = https://github.com/ESCOMP/CDEPS.git
local_path = components/cdeps
required = True

[fv3gfs_interface]
# ufs_update_11
tag = 6a78103
protocol = git
repo_url = https://github.com/ESCOMP/fv3gfs_interface.git
local_path = HAFS/sorc/hafs_forecast.fd/FV3/cime/
required = True

[nems_interface]
tag = 3c36830
protocol = git
repo_url = https://github.com/ESCOMP/NEMS_interface.git
local_path = HAFS/sorc/hafs_forecast.fd/NEMS/cime/
required = True
```

but I am getting error like following,

```
Processing externals description file : Externals.cfg
Checking status of externals: hycom, cime, fv3gfs_interface, hafs, cdeps, nems_interface, 
Checking out externals: hycom, cime, fv3gfs_interface, hafs, ERROR:root:Command '[u'git', u'checkout', u'--quiet', u'hafs-community_HAFS.git/feature/hafs_nogsi']' returned non-zero exit status 1
ERROR:root:Failed with output:
    error: Your local changes to the following files would be overwritten by checkout:
    	Externals.cfg
    Please commit your changes or stash them before you switch branches.
    Aborting

ERROR: In directory
    /glade/u/home/turuncu/UFS/ufs-hafs-app/HAFS
Process did not run successfully; returned status 1:
    git checkout --quiet hafs-community_HAFS.git/feature/hafs_nogsi
See above for output from failed command.

ERROR:root:Failed with output:
    error: Your local changes to the following files would be overwritten by checkout:
    	Externals.cfg
    Please commit your changes or stash them before you switch branches.
    Aborting
```

I think that the reason of getting this error is to checkout  `fv3gfs_interface` before checking out `hafs` because  `fv3gfs_interface` is checking out as a subfolder in HAFs model. Currently, there is no way to control the order of checking out externals in the order that are defined in the Externals.cfg file and that could be a nice feature. I try to look at the possible solution and it seems that the configuration file is read by **ConfigParser** module and I could not find any information to read with that way. I tried to sort the models but that also not working because the alphabetical order also causes same problem. So, I need explicit control of checking out externals. Any suggestion will be helpful at this point.

### PR DESCRIPTION
**Pull request by jedwards4b**
Add a prereq optional field to the Externals file.

Adds an optional prereq field to an external description.  This field if present should be the name of another repository
in the same file.  

User interface changes?: 
[ If yes, describe what changed, and steps taken to ensure backward compatibilty ]

Fixes: #147 

Testing:
  test removed: None
  unit tests: All pass
  system tests: All pass
  manual testing: ufs-s2s-app

### CODE REVIEW

File: manic/sourcetree.py [hunk 1]
```diff
@@ -331,11 +339,23 @@ def checkout(self, verbosity, load_all, load_comp=None):
...
+            tmp_comps = self._required_compnames
+
+        load_comps = []
+        for comp in tmp_comps:
+            prereq = self._all_components[comp].get_prereq()
+            if prereq:
+                if prereq in self._all_components and prereq not in load_comps:
+                    load_comps.append(prereq)
+                else:
+                    fatal_error("prereq {} of component {} not found".format(prereq, comp))
```
Comment (billsacks) [inline manic/sourcetree.py]: This `else` can be hit in two cases:

1. if prereq is not in self._all_components

2. if prereq is already in load_comps

If I understand this correctly, the error should only be raised due to condition (1), not due to condition (2)

File: manic/sourcetree.py [hunk 2]
```diff
@@ -331,11 +339,23 @@ def checkout(self, verbosity, load_all, load_comp=None):
...
         else:
-            load_comps = self._required_compnames
+            tmp_comps = self._required_compnames
+
+        load_comps = []
+        for comp in tmp_comps:
+            prereq = self._all_components[comp].get_prereq()
+            if prereq:
+                if prereq in self._all_components and prereq not in load_comps:
+                    load_comps.append(prereq)
```
Comment (billsacks) [inline manic/sourcetree.py]: Unless I'm missing something, this doesn't check if the given prereq itself has prerequisites. I'm thinking this should be checked recursively. In doing so, there should probably be an error raised if a circularity of prerequisites is found.

**Review by billsacks — CHANGES_REQUESTED** [review]
Thanks for working on this @jedwards4b . I've just looked at the core logic of the new feature; I have a couple of comments.

**Comment by jedwards4b** [comment]
@billsacks thank you for reviewing, I think that I have addressed the issues you've raised.

File: test/.pylint.rc [hunk 3]
```diff
@@ -50,7 +50,7 @@ confidence=
 # --enable=similarities". If you want to run only the classes checker, but have
 # no Warning level messages displayed, use"--disable=all --enable=classes
 # --disable=W"
-disable=bad-continuation,useless-object-inheritance
+disable=bad-continuation,useless-object-inheritance,bad-option-value
```
Comment (gold2718) [inline test/.pylint.rc:53]: What is this for? What 'bad' option value is not bad in your opinion?

**Comment by gold2718** [comment]
I think #150 (bug fix) should go in before this enhancement.

**Review by gold2718 — CHANGES_REQUESTED** [review]
I have not looked at sourcetree in detail yet but with no tests, it is impossible to know if this does what it is supposed to (never mind protecting that functionality against future changes). Some things that do not appear to work:

- Repo A is a `prereq` for repo B but you put `prereq = B` in repo A: (B is checked out but not A).
- Invalid `prereq` name: (circular dependency is reported along with a not found error even if no circular dependency)
- Repo A is a `prereq` for B but A is marked `required = False`: (B is not checked out)
- Everything is correct but sort (hash order) of `prereq` (A) is after dependent (B): B is not checked out

Add some tests for correct functionality and error handling and then fix all these bugs.

File: test/.pylint.rc (hunk 3 above)
Comment (jedwards4b) [reply test/.pylint.rc:53]: bad-option-value is for # pylint: disable=R0204 redefined-variable-type
which is only reported in py3.5 or newer.   I could probably use some help in getting the
tests to pass, it seems like I am responsible for fixing things that I did not break.

**Comment by jedwards4b** [comment]
> I have not looked at sourcetree in detail yet but with no tests, it is impossible to know if this does what it is supposed to (never mind protecting that functionality against future changes). Some things that do not appear to work:
> 
>     * Repo A is a `prereq` for repo B but you put `prereq = B` in repo A: (B is checked out but not A).
> 
Can you please elaborate?  Are you asking me to recognize a prereq relationship that is not spelled out (or incorrectly set) in the config file, trap it and report the error?  I think that may be beyond the scope of this PR. 

>     * Invalid `prereq` name: (circular dependency is reported along with a not found error even if no circular dependency)
> 
Fixed


>     * Repo A is a `prereq` for B but A is marked `required = False`: (B is not checked out)
> 
This is not true, in this case A and B are both checked out. 


>     * Everything is correct but sort (hash order) of `prereq` (A) is after dependent (B): B is not checked out
> 
Also not true, did you try it?  This is the entire point of the PR.


> 
> Add some tests for correct functionality and error handling and then fix all these bugs.

Having a hard time understanding the test structure, I see for example  test_container_simple_required but I don't see where it is called.

**Comment by jedwards4b** [comment]
I've added a unit test, that *I think* just makes sure that a prereq can be written to and read from the externals.cfg file.
I would like to add a test for invalid prereq and for circular dependencies, but I haven't figured out how to do that yet.
I would like to start with a simple test - if component A includes a prereq which is not a comp in the file throw an error.
I think if you @gold2718 could help with one, I may grasp enough to write more on my own.

**Comment by gold2718** [comment]
> Having a hard time understanding the test structure, I see for example test_container_simple_required but I don't see where it is called.

It's just standard python unittest usage, what is confusing about this case?

**Comment by gold2718** [comment]
> Also not true, did you try it? This is the entire point of the PR.

Yes, that is why I mentioned it. I noticed that the hash order of my `prereq` repo was before the `prereq` user so I fiddled with the name until the hash order switched. I think there should be a test like that because _that_ is the "entire point of the PR". If the hash order already meets the `prereq` requirements, no code changes are required.

**Comment by gold2718** [comment]
Here is an externals file I think does not work:
```
[cmeps]
hash = 471ac52
protocol = git
repo_url = https://github.com/ESCOMP/CMEPS.git
local_path = cime/src/drivers/nuopc/
required = True
prereq = xcimex

[xcimex]
tag = cime5.8.24
protocol = git
repo_url = https://github.com/ESMCI/cime
local_path = cime
required = True

[externals_description]
schema_version = 1.0.01
```
Some output:
```
$ ./checkout_externals --status
Processing externals description file : Externals.cfg
Checking status of externals: cmeps, xcimex, 
    ./cime
e-  ./cime/src/drivers/nuopc/
$ ./checkout_externals
Processing externals description file : Externals.cfg
Checking status of externals: cmeps, xcimex, 
Checking out externals: load_comps ['xcimex']
xcimex, 

$ ./checkout_externals --status
Processing externals description file : Externals.cfg
Checking status of externals: cmeps, xcimex, 
    ./cime
e-  ./cime/src/drivers/nuopc/
```

**Comment by jedwards4b** [comment]
Using the above and python 2.7.13 it works fine. However I am able to reproduce your result with 3.7.5, I am looking into it. 
Can we put this into a test?   Which test file would be most appropriate?

**Review by billsacks — APPROVED** [review]
Thanks for your changes, @jedwards4b . I'm marking this as approved because you have addressed my comments - thanks. I haven't looked closely at all of @gold2718 's comments, and I will defer to him on final approval of this PR - but I can also look more closely if one of you want me to.

I do see two possible issues with this implementation:

1. It only handles a single prerequisite for each component

2. Relying on python's fail-safe recursion limit to detect circular dependencies feels a little wrong to me, but maybe it's okay

To address these issues, we could use an already-developed topological sort routine. I found one here that looks like it does what we'd want (possibly with a bit of packing / unpacking of the data before and after to put it in the right data structures): https://pypi.org/project/toposort. The implementation is simple enough that we could probably just inline this into manage_externals (as long as we preserve the license). (I also found what looks like a similar solution on stack overflow, but I haven't looked at it as closely as I looked at the above toposort: https://stackoverflow.com/questions/11557241/python-sorting-a-dependency-list .)

I don't feel that is critical, though, so I'm marking this as approved, and will leave it up to you whether it's worth adopting a solution like that now, or leaving that for later.

**Pull request closed without merging**